# 분석을 깊게 하고 싶으면 R을 해야해요
- 근데 왜 우리는 R을 안 쓸까?
    `- 장점이 지금은 하나뿐임.
     -  R 통계적 새로운 기법이 같이 딸려서 나와. 따라서 바로 적용이 가능해

# 추천시스템 Recommendation
- 정보를 찾는 방법
    - 1. 검색: 적극적. 특정 쿼리(검색어)를 통해 특정하여 찾아주는 방식
    - 2. 추천: 수동적. 원하는 정보를 과거 데이터에 기반하여 미래를 예측하여줌
        - 추천은 기계학습과 별도로 발전되었으나 어느순간부터 추천==기계학습이 되었어요.
        
#### ■파레토법칙
 - 소품종 대량생산: 선택과 집중

#### ■롱테일의 법칙
- 다품종 소량생산(Niche market_니시의 마켓). 대중의 힘

#### ■필터버블
- 필터가 거품처럼 늘어남
- 생각조종자들이 번역
    - 페이스북) 특정 페이스북 보는 사람들은 그쪽으로 생각이 편향
    - 이걸 이용하면 감정 변경 및 여론 조장이 가능해짐
    
- "스마트한 생각들"
- "스마트한 선택들"
- "조지오웬 1984"- 텔레스크린, 빅브라더 개념. 
- "멋진 신세계" 

### 추천 시스템 종류2가지
#### - 협업 필터링(사람 기반)
    - 나랑 유사한 사람에게 유사한 아이템을 추천할 확률이 높아
    - 같은 동네/같은 성별이면 행동도 비슷할 것이다가 추천 시스템의 핵심
#### - 콘탠츠 기반 필터링(아이템 기반)
    - 항상 클래식을 들으면 앞으로도 클래식 들을 거야
    - 이전 콘탠츠와의 유사도를 비교하여 그와 유사한 것을 내게 추천

#### 비슷한 패턴을 사용하는 항목이나 사용자를 추출하는게 핵심
- 사용자 끼리 유사도 찾는것이 핵심
- 그럼 유사도를 어떻게 찾을까?
    - 상점 입장)구매내역 등 과거의 패턴 기반 유사도 체크

#### serendipity
- 불확실성인데 이걸 추천시스템에 녹인다니. 정말 매력적이래
- 신선한 추천
- 나랑 유사한 사람이 나랑 같은거 다했는데, 하나만 달라. 그러면 그 안 한 한개를 추천해줄 수도 있지
- 물건 파는 사람들이 좋아하고, 이게 대부분의 쇼핑몰에서 지원하는 추천방식이야
---------------------------------------
- 과거 구매 내역등을 기반으로 추천을 들어가서 우리는 시스템 만들기 좋고 성능도 좋아

### 항상 밝은 면만 있는건 아니겠죠
#### - 문제점
    - 1. 콜드 스타트 문제
        - 새로운 물건에 대해서는 과거 데이터가 없죠. 
        - 힘들고 추운 상태에서 시작해. 
        - 새로운거 시작할때 성능이 안 좋았던 이유가 바로 이떄문. 
    - 2. 롱테일 문제
        - 항상 사람들에 많이 팔린건 유사도 높아서 비대칭적 쏠림이 일반적.
        - 이거 해결하려고 했는데 20프로의 또 20프로에 쏠리는 식
    - 3. gray sheep/black sheep 문제
        - 독특한 취향의 문제
        - 왔다갔다 애매한 사람. 혹은 선호도가 이상한 사람

### 콘텐츠 기반 필터링
- 이게 꽤 어려워요
- 핵심) 내용에서 어떻게 유사도를 찾는가

## 하이브리드 시스템
- 협업 필터링 & 콘텐츠 기반 필터링 합친 것

## 모델 기반 협력 필터링
- 기계학습 모델
    - 유사도: 기계학습기법. KNN+숨어있는 의미 찾는 방식으로 가는 중 
        - 상관계수/스피어만/켄달/피어슨
    - 주성분, SVD
    

### 오늘은 판다스로 메모리 베이스 접근 직접 짜볼거야

### 기본적인 데이터 생성
1. 과거 데이터 입력
    - 아이템 어떤거 구매했다. 
2. 숫자로 변경
    - 구매내역1, 안 구매 내역0
    - 이걸 바이너리 분류라고 해요.
    - 샀는지 유무를 따지면 0,1
    - 물품은 동일한거 여러개 잘 안 사. 
3. 혹은 평점도 입력할 수 있어★
    - 그 외에 비어있는건 0으로 기입


#### 기계학습은 피쳐/디맨션이라고 하는 이 컬럼을 
- 즉 가로는 피쳐, 세로 내역들은 벡터. 
- 이거 두 개 합쳐서 메트릭스 만들거에요.
- 이게 기본 준비 상태

### 사람끼리 유사도 찾는 방법?
- 코사인 시뮬러리티. 

#### 내게 제일 가까운애 차장
- 찾아서 소팅
- 그 이웃이 구매한 내역 나열
- 내가 샀던거를 그 중에서 제외해
- 유사도 가장 큰 사람이 내가 안 샀던 물건을 바로 추천해줄수도 있고
- 값 비교
    - 가격까지 비교해서 추천
 
 #### 질문) KNN은 K 홀수인 경우인데 추천 시스템은 고려 안 하나요?
 - 내게서 유사한 거리에 있는 몇개를 보고 추천하는데 짝수면 동점일 경우 곤란하지. 그래서 홀수 쓰면 좋단거야.

In [4]:
# 내가 샀던거 기반으로 유사도 측정. 열을 변경해서 했더니 

# 아이템 베이스 콜라보레이트
- 콘텐츠 속성 고려하지 ㅇ낳고
- 사람들이 과거에 어떤거 했는지를 보고 유사도를 측정
- 디비에 저장 하는등 하면서 캐시로 빠르게 사용 가능+ 선응도 좋았다. 

# 콘텐츠 베이스 콜라보레이티브
- 메타데이터(콘텐츠 고유의 속성) 만으로 유사도 고려하여 테이블 생성
- 미술 관련된 책 추천해준다면

#### 콜라보레이티브 차이
1. 사람이 한 과거의 행동(아이템기반/유저기반)
2. 컨텐츠 베이스는 행동과 상관없이 콘탠츠고유의 속성만 가지고 유사도를 계산


# KNN 
- 알고리즘적 구조적으로 안 어려워
- 문제는 검증을 어떻게 할 건지.
    - 그래서 크리티컬한 분야에서는 예측을 잘 안 써
    - 네플릭스 추천과 곤련해서는 사람가지고 체크를 해. 
        - 추천 받은 사람이 내 마음에 드네. 하게끔 검증을 했어요 거꾸로.
- 알고리즘 여러개 비교해서 사람 대상으로 검증을 했어요. 
- 왜 추천시스템 만들게 하냐면 입사할떄 아주 좋아요. 검증이슈도 줄고 구현되고 회사에서 요구하고

## 추천을 어디에 넣는가도 이슈
- 대체재와 보완재
- 핸드폰 구매하면서 그 옆에 핸드폰 보재액세서리 추천하는식.

## 연습용 데이터를 받아볼게요
- movielens
    - 영화 평점 사이트
    - 어느정도 타이디 데이터 형식으로 저장이 되어 있어


## 1.데이터 불러와야해요
- 메모장으로 열린다
        - 텍스트 형식으로 저장이 되었다.
        - 확장자는 중요하지 않아.

### u.data, u.genre, u.info, u.item, u.occupation, u.user 불러봐요.
- read_csv는 ~@가 아니어도 불러들일수있었죠. 

In [7]:
import pandas as pd

In [8]:
#u.data
data=pd.read_csv('C:/Python/ml-100k/ml-100k/u.data', engine='python', header=None, sep='\t')

In [9]:
data

0     1  2          3
0      196   242  3  881250949
1      186   302  3  891717742
2       22   377  1  878887116
3      244    51  2  880606923
4      166   346  1  886397596
5      298   474  4  884182806
6      115   265  2  881171488
7      253   465  5  891628467
8      305   451  3  886324817
9        6    86  3  883603013
10      62   257  2  879372434
11     286  1014  5  879781125
12     200   222  5  876042340
13     210    40  3  891035994
14     224    29  3  888104457
15     303   785  3  879485318
16     122   387  5  879270459
17     194   274  2  879539794
18     291  1042  4  874834944
19     234  1184  2  892079237
20     119   392  4  886176814
21     167   486  4  892738452
22     299   144  4  877881320
23     291   118  2  874833878
24     308     1  4  887736532
25      95   546  2  879196566
26      38    95  5  892430094
27     102   768  2  883748450
28      63   277  4  875747401
29     160   234  5  876861185
...    ...   ... ..        ...
99970  449   120  1  879959573
99971  661   762  2  876037121
99972  721   874  3  877137447
99973  821   151  4  874792889
99974  764   596  3  876243046
99975  537   443  3  886031752
99976  618   628  2  891308019
99977  487   291  3  883445079
99978  113   975  5  875936424
99979  943   391  2  888640291
99980  864   685  4  888891900
99981  750   323  3  879445877
99982  279    64  1  875308510
99983  646   750  3  888528902
99984  654   370  2  887863914
99985  617   582  4  883789294
99986  913   690  3  880824288
99987  660   229  2  891406212
99988  421   498  4  892241344
99989  495  1091  4  888637503
99990  806   421  4  882388897
99991  676   538  4  892685437
99992  721   262  3  877137285
99993  913   209  2  881367150
99994  378    78  3  880056976
99995  880   476  3  880175444
99996  716   204  5  879795543
99997  276  1090  1  874795795
99998   13   225  2  882399156
99999   12   203  3  879959583

[100000 rows x 4 columns]

In [10]:
data.rename({0:'user_id', 1:'item_id', 2:'rating', 3:'timestamp'}, axis=1, inplace=True)

In [11]:
#u.genre
genre=pd.read_csv('C:\Python\ml-100k\ml-100k/u.genre', engine='python', header=None)

#### 효율성을 위해 장르 인덱스 없애는게 좋을 듯 

In [12]:
genre

0
0       unknown|0
1        Action|1
2     Adventure|2
3     Animation|3
4    Children's|4
5        Comedy|5
6         Crime|6
7   Documentary|7
8         Drama|8
9       Fantasy|9
10   Film-Noir|10
11      Horror|11
12     Musical|12
13     Mystery|13
14     Romance|14
15      Sci-Fi|15
16    Thriller|16
17         War|17
18     Western|18

In [ ]:
#### 장르 짝대기도 없고, 헤더도 없고 

In [13]:
#u.info
info=pd.read_csv('C:/Python/ml-100k/ml-100k/u.info', header=None, sep=' ')

### 잘열리는듯 해보이지만 아니라서, headr=None을 써줘야해
- 2개가 아니고 여러개가 있었어야하지.
- sep=' ' 공백으로도 분리시켰어
- 그러니 이제 데이터가 더 잘 보이네
- 유저 레이팅의 수래요. 


In [19]:
info

0        1
0     943    users
1    1682    items
2  100000  ratings

####  이름 바꿔볼까

In [14]:
info.rename({1:'table', 0:'number'}, axis=1)

number    table
0     943    users
1    1682    items
2  100000  ratings

In [15]:
#u.item
item= pd.read_csv('C:/Python/ml-100k/ml-100k/u.item', header=None,sep=' ')

ParserError: Error tokenizing data. C error: Expected 3 fields in line 6, saw 10


#### 아이템 이거 engine='python'

In [16]:
item= pd.read_csv('C:/Python/ml-100k/ml-100k/u.item', header=None,sep=' ', engine='python')

ParserError: Expected 3 fields in line 6, saw 10

#### item을 텍스트로 열려고 하니 '유니코드'로 안 열린다고 창이 뜨네
- 영화명에 외국어 명이 들 수 있어
- 그래서 encoding='latin1 라틴을 했어
    - CSV는 기본적으로 콤마로 나눠지는데 여기서 12번째 행 열면 'Usual Suspects, The'하면서 그 안에 콤마 있어서 안 열린거야

#### csv 
- csv는 기본적으로 콤마를 체크해요
- 그래서 sep를 짝대기 해서 

In [17]:
item= pd.read_csv('C:/Python/ml-100k/ml-100k/u.item', sep='|', header=None, engine='python', encoding='latin1')

In [ ]:
#item= pd.read_csv('C:/Python/ml-100k/ml-100k/u.item', delemiter='|', header=None, engine='python', encoding='latin1')

In [18]:
item

0                                                  1            2   \
0        1                                   Toy Story (1995)  01-Jan-1995   
1        2                                   GoldenEye (1995)  01-Jan-1995   
2        3                                  Four Rooms (1995)  01-Jan-1995   
3        4                                  Get Shorty (1995)  01-Jan-1995   
4        5                                     Copycat (1995)  01-Jan-1995   
5        6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...  01-Jan-1995   
6        7                              Twelve Monkeys (1995)  01-Jan-1995   
7        8                                        Babe (1995)  01-Jan-1995   
8        9                            Dead Man Walking (1995)  01-Jan-1995   
9       10                                 Richard III (1995)  22-Jan-1996   
10      11                               Seven (Se7en) (1995)  01-Jan-1995   
11      12                         Usual Suspects, The (1995)  14-Aug-1995   
12      13                            Mighty Aphrodite (1995)  30-Oct-1995   
13      14                                 Postino, Il (1994)  01-Jan-1994   
14      15                          Mr. Holland's Opus (1995)  29-Jan-1996   
15      16                 French Twist (Gazon maudit) (1995)  01-Jan-1995   
16      17                         From Dusk Till Dawn (1996)  05-Feb-1996   
17      18                          White Balloon, The (1995)  01-Jan-1995   
18      19                              Antonia's Line (1995)  01-Jan-1995   
19      20                          Angels and Insects (1995)  01-Jan-1995   
20      21                      Muppet Treasure Island (1996)  16-Feb-1996   
21      22                                  Braveheart (1995)  16-Feb-1996   
22      23                                 Taxi Driver (1976)  16-Feb-1996   
23      24                         Rumble in the Bronx (1995)  23-Feb-1996   
24      25                               Birdcage, The (1996)  08-Mar-1996   
25      26                      Brothers McMullen, The (1995)  01-Jan-1995   
26      27                                    Bad Boys (1995)  01-Jan-1995   
27      28                                   Apollo 13 (1995)  01-Jan-1995   
28      29                              Batman Forever (1995)  01-Jan-1995   
29      30                               Belle de jour (1967)  01-Jan-1967   
...    ...                                                ...          ...   
1652  1653  Entertaining Angels: The Dorothy Day Story (1996)  27-Sep-1996   
1653  1654                       Chairman of the Board (1998)  01-Jan-1998   
1654  1655                                  Favor, The (1994)  01-Jan-1994   
1655  1656                                 Little City (1998)  20-Feb-1998   
1656  1657                                      Target (1995)  28-Feb-1996   
1657  1658                      Substance of Fire, The (1996)  06-Dec-1996   
1658  1659                    Getting Away With Murder (1996)  12-Apr-1996   
1659  1660                                 Small Faces (1995)  09-Aug-1996   
1660  1661                                New Age, The (1994)  01-Jan-1994   
1661  1662                                 Rough Magic (1995)  30-May-1997   
1662  1663                            Nothing Personal (1995)  30-Apr-1997   
1663  1664                     8 Heads in a Duffel Bag (1997)  18-Apr-1997   
1664  1665                           Brother's Kiss, A (1997)  25-Apr-1997   
1665  1666                                        Ripe (1996)  02-May-1997   
1666  1667                              Next Step, The (1995)  13-Jun-1997   
1667  1668                          Wedding Bell Blues (1996)  13-Jun-1997   
1668  1669                           MURDER and murder (1996)  20-Jun-1997   
1669  1670                                     Tainted (1998)  01-Feb-1998   
1670  1671                          Further Gesture, A (1996)  20-Feb-1998   
1671  1672                    

In [19]:
item.rename({1:'table', 0:'number'}, axis=1)

number                                              table            2  \
0          1                                   Toy Story (1995)  01-Jan-1995   
1          2                                   GoldenEye (1995)  01-Jan-1995   
2          3                                  Four Rooms (1995)  01-Jan-1995   
3          4                                  Get Shorty (1995)  01-Jan-1995   
4          5                                     Copycat (1995)  01-Jan-1995   
5          6  Shanghai Triad (Yao a yao yao dao waipo qiao) ...  01-Jan-1995   
6          7                              Twelve Monkeys (1995)  01-Jan-1995   
7          8                                        Babe (1995)  01-Jan-1995   
8          9                            Dead Man Walking (1995)  01-Jan-1995   
9         10                                 Richard III (1995)  22-Jan-1996   
10        11                               Seven (Se7en) (1995)  01-Jan-1995   
11        12                         Usual Suspects, The (1995)  14-Aug-1995   
12        13                            Mighty Aphrodite (1995)  30-Oct-1995   
13        14                                 Postino, Il (1994)  01-Jan-1994   
14        15                          Mr. Holland's Opus (1995)  29-Jan-1996   
15        16                 French Twist (Gazon maudit) (1995)  01-Jan-1995   
16        17                         From Dusk Till Dawn (1996)  05-Feb-1996   
17        18                          White Balloon, The (1995)  01-Jan-1995   
18        19                              Antonia's Line (1995)  01-Jan-1995   
19        20                          Angels and Insects (1995)  01-Jan-1995   
20        21                      Muppet Treasure Island (1996)  16-Feb-1996   
21        22                                  Braveheart (1995)  16-Feb-1996   
22        23                                 Taxi Driver (1976)  16-Feb-1996   
23        24                         Rumble in the Bronx (1995)  23-Feb-1996   
24        25                               Birdcage, The (1996)  08-Mar-1996   
25        26                      Brothers McMullen, The (1995)  01-Jan-1995   
26        27                                    Bad Boys (1995)  01-Jan-1995   
27        28                                   Apollo 13 (1995)  01-Jan-1995   
28        29                              Batman Forever (1995)  01-Jan-1995   
29        30                               Belle de jour (1967)  01-Jan-1967   
...      ...                                                ...          ...   
1652    1653  Entertaining Angels: The Dorothy Day Story (1996)  27-Sep-1996   
1653    1654                       Chairman of the Board (1998)  01-Jan-1998   
1654    1655                                  Favor, The (1994)  01-Jan-1994   
1655    1656                                 Little City (1998)  20-Feb-1998   
1656    1657                                      Target (1995)  28-Feb-1996   
1657    1658                      Substance of Fire, The (1996)  06-Dec-1996   
1658    1659                    Getting Away With Murder (1996)  12-Apr-1996   
1659    1660                                 Small Faces (1995)  09-Aug-1996   
1660    1661                                New Age, The (1994)  01-Jan-1994   
1661    1662                                 Rough Magic (1995)  30-May-1997   
1662    1663                            Nothing Personal (1995)  30-Apr-1997   
1663    1664                     8 Heads in a Duffel Bag (1997)  18-Apr-1997   
1664    1665                           Brother's Kiss, A (1997)  25-Apr-1997   
1665    1666                                        Ripe (1996)  02-May-1997   
1666    1667                              Next Step, The (1995)  13-Jun-1997   
1667    1668                          Wedding Bell Blues (1996)  13-Jun-1997   
1668    1669                           MURDER and murder (1996)  20-Jun-1997   
1669    1670                                     Tainted (1998)  01-Feb-1998   
1670  

#### item에 대한 이름 변경(readme 참고해서)

In [20]:
item.rename({0:'movie_id',1:'title',2:'release',3:'video release',4:'IMDb URL', 5:'unknown',6:'Action',7:'Adventure',8:'Animation',9:"Children's",10:'Comedy',11:'Crime',12:'Documentary',13:'Drama',14:'Fantasy',15:'Film-Noir',16:'Horror',17:'Musical',18:'Mystery',19:'Romance',20:'Sci-Fi',21:'Thriller',22:'War',23:'Western'},axis=1, inplace=True)

#### item은 타이드 데이터가 아니네요~! 바꿔야해◆
- 이름 바꾸기 전에 멜트하면 더 우아했겠죠
- unknown부터 western까지 ㅠ
- 장르 이놈도 바보 데이터래
- 근데 잘 보면 
## melt 해줍시다
- 근데 장르가 로맨스/스릴러 2개 이상 겹칠 수 있으니까

In [21]:
item.columns[:5]

Index(['movie_id', 'title', 'release', 'video release', 'IMDb URL'], dtype='object')

In [23]:
item=item.melt(item.columns[:5])

In [24]:
len(item['id']==1)

KeyError: 'id'

In [25]:
item[item['id']==1]

KeyError: 'id'

# 우리는 추천 시스템 만들떄 처음에 할일?
- 테이블 만들기
- 테이블의 구성요소는 몇 개? 
    - 왼쪽에 유저, 오른쪽에는 영화 제목, 그리고 가운데 값에는 영화 평점이 들어가야해
    - 이런 정보 들어간 테이블이 data 테이블 
        - user, item, rating있네요. 
        - 이거 ㄹ이용하면 필요한 행렬 만들 수 있어
            - userid를 왼쪽에 넘기는건? 인덱스 

### 재미삼아 머지 해볼게요

In [26]:
data.merge(user)

NameError: name 'user' is not defined

#### merge
- 분리된 테이블 inner join

#### 조인
- 컬럼끼리 합칠때

### pd.to_datetime(data.timestamp)
- 시간 데이터 변경
- timestamp는 1970년1월1일부터 플러스 일
- 도움말 보면, unit있어

In [90]:
pd.to_datetime(data.timestamp)

0       1970-01-01 00:00:00.881250949
1       1970-01-01 00:00:00.891717742
2       1970-01-01 00:00:00.878887116
3       1970-01-01 00:00:00.880606923
4       1970-01-01 00:00:00.886397596
5       1970-01-01 00:00:00.884182806
6       1970-01-01 00:00:00.881171488
7       1970-01-01 00:00:00.891628467
8       1970-01-01 00:00:00.886324817
9       1970-01-01 00:00:00.883603013
10      1970-01-01 00:00:00.879372434
11      1970-01-01 00:00:00.879781125
12      1970-01-01 00:00:00.876042340
13      1970-01-01 00:00:00.891035994
14      1970-01-01 00:00:00.888104457
15      1970-01-01 00:00:00.879485318
16      1970-01-01 00:00:00.879270459
17      1970-01-01 00:00:00.879539794
18      1970-01-01 00:00:00.874834944
19      1970-01-01 00:00:00.892079237
20      1970-01-01 00:00:00.886176814
21      1970-01-01 00:00:00.892738452
22      1970-01-01 00:00:00.877881320
23      1970-01-01 00:00:00.874833878
24      1970-01-01 00:00:00.887736532
25      1970-01-01 00:00:00.879196566
26      1970

In [91]:
pd.to_datetime(data.timestamp, unit='s') #판다스 나노 세컨드인데, 이거 세컨드로 바꿨음

0       1997-12-04 15:55:49
1       1998-04-04 19:22:22
2       1997-11-07 07:18:36
3       1997-11-27 05:02:03
4       1998-02-02 05:33:16
5       1998-01-07 14:20:06
6       1997-12-03 17:51:28
7       1998-04-03 18:34:27
8       1998-02-01 09:20:17
9       1997-12-31 21:16:53
10      1997-11-12 22:07:14
11      1997-11-17 15:38:45
12      1997-10-05 09:05:40
13      1998-03-27 21:59:54
14      1998-02-21 23:40:57
15      1997-11-14 05:28:38
16      1997-11-11 17:47:39
17      1997-11-14 20:36:34
18      1997-09-21 09:42:24
19      1998-04-08 23:47:17
20      1998-01-30 16:13:34
21      1998-04-16 14:54:12
22      1997-10-26 15:55:20
23      1997-09-21 09:24:38
24      1998-02-17 17:28:52
25      1997-11-10 21:16:06
26      1998-04-13 01:14:54
27      1998-01-02 13:40:50
28      1997-10-01 23:10:01
29      1997-10-14 20:33:05
                ...        
99970   1997-11-19 17:12:53
99971   1997-10-05 07:38:41
99972   1997-10-18 01:17:27
99973   1997-09-20 22:01:29
99974   1997-10-07 1

#### 원본

In [92]:
data_raw= data.copy()

In [104]:
data_raw

user_id  item_id  rating  timestamp
0          196      242       3  881250949
1          186      302       3  891717742
2           22      377       1  878887116
3          244       51       2  880606923
4          166      346       1  886397596
5          298      474       4  884182806
6          115      265       2  881171488
7          253      465       5  891628467
8          305      451       3  886324817
9            6       86       3  883603013
10          62      257       2  879372434
11         286     1014       5  879781125
12         200      222       5  876042340
13         210       40       3  891035994
14         224       29       3  888104457
15         303      785       3  879485318
16         122      387       5  879270459
17         194      274       2  879539794
18         291     1042       4  874834944
19         234     1184       2  892079237
20         119      392       4  886176814
21         167      486       4  892738452
22         299      144       4  877881320
23         291      118       2  874833878
24         308        1       4  887736532
25          95      546       2  879196566
26          38       95       5  892430094
27         102      768       2  883748450
28          63      277       4  875747401
29         160      234       5  876861185
...        ...      ...     ...        ...
99970      449      120       1  879959573
99971      661      762       2  876037121
99972      721      874       3  877137447
99973      821      151       4  874792889
99974      764      596       3  876243046
99975      537      443       3  886031752
99976      618      628       2  891308019
99977      487      291       3  883445079
99978      113      975       5  875936424
99979      943      391       2  888640291
99980      864      685       4  888891900
99981      750      323       3  879445877
99982      279       64       1  875308510
99983      646      750       3  888528902
99984      654      370       2  887863914
99985      617      582       4  883789294
99986      913      690       3  880824288
99987      660      229       2  891406212
99988      421      498       4  892241344
99989      495     1091       4  888637503
99990      806      421       4  882388897
99991      676      538       4  892685437
99992      721      262       3  877137285
99993      913      209       2  881367150
99994      378       78       3  880056976
99995      880      476       3  880175444
99996      716      204       5  879795543
99997      276     1090       1  874795795
99998       13      225       2  882399156
99999       12      203       3  879959583

[100000 rows x 4 columns]

#### 필요없는 타임 스탬프 지우기
- drop

In [94]:
data.drop(columns='timestamp', inplace=True)

In [95]:
data

user_id  item_id  rating
0          196      242       3
1          186      302       3
2           22      377       1
3          244       51       2
4          166      346       1
5          298      474       4
6          115      265       2
7          253      465       5
8          305      451       3
9            6       86       3
10          62      257       2
11         286     1014       5
12         200      222       5
13         210       40       3
14         224       29       3
15         303      785       3
16         122      387       5
17         194      274       2
18         291     1042       4
19         234     1184       2
20         119      392       4
21         167      486       4
22         299      144       4
23         291      118       2
24         308        1       4
25          95      546       2
26          38       95       5
27         102      768       2
28          63      277       4
29         160      234       5
...        ...      ...     ...
99970      449      120       1
99971      661      762       2
99972      721      874       3
99973      821      151       4
99974      764      596       3
99975      537      443       3
99976      618      628       2
99977      487      291       3
99978      113      975       5
99979      943      391       2
99980      864      685       4
99981      750      323       3
99982      279       64       1
99983      646      750       3
99984      654      370       2
99985      617      582       4
99986      913      690       3
99987      660      229       2
99988      421      498       4
99989      495     1091       4
99990      806      421       4
99991      676      538       4
99992      721      262       3
99993      913      209       2
99994      378       78       3
99995      880      476       3
99996      716      204       5
99997      276     1090       1
99998       13      225       2
99999       12      203       3

[100000 rows x 3 columns]

#### userID를 인덱스로 옮겨요
- setindex

In [96]:
data.set_index('user_id')

item_id  rating
user_id                 
196          242       3
186          302       3
22           377       1
244           51       2
166          346       1
298          474       4
115          265       2
253          465       5
305          451       3
6             86       3
62           257       2
286         1014       5
200          222       5
210           40       3
224           29       3
303          785       3
122          387       5
194          274       2
291         1042       4
234         1184       2
119          392       4
167          486       4
299          144       4
291          118       2
308            1       4
95           546       2
38            95       5
102          768       2
63           277       4
160          234       5
...          ...     ...
449          120       1
661          762       2
721          874       3
821          151       4
764          596       3
537          443       3
618          628       2
487          291       3
113          975       5
943          391       2
864          685       4
750          323       3
279           64       1
646          750       3
654          370       2
617          582       4
913          690       3
660          229       2
421          498       4
495         1091       4
806          421       4
676          538       4
721          262       3
913          209       2
378           78       3
880          476       3
716          204       5
276         1090       1
13           225       2
12           203       3

[100000 rows x 2 columns]

#### 이제 저 유저 아이디를 어떻게 펼칠까
- SCI
- 인덱스 두개 넣어볼게
- unstack하고 setindex 하는 것과 피봇이 동일한 결과를 보여주는데 피봇이 교수님이 더 이쁜거 같다셔

In [98]:
data.set_index(['user_id', 'item_id'])

rating
user_id item_id        
196     242           3
186     302           3
22      377           1
244     51            2
166     346           1
298     474           4
115     265           2
253     465           5
305     451           3
6       86            3
62      257           2
286     1014          5
200     222           5
210     40            3
224     29            3
303     785           3
122     387           5
194     274           2
291     1042          4
234     1184          2
119     392           4
167     486           4
299     144           4
291     118           2
308     1             4
95      546           2
38      95            5
102     768           2
63      277           4
160     234           5
...                 ...
449     120           1
661     762           2
721     874           3
821     151           4
764     596           3
537     443           3
618     628           2
487     291           3
113     975           5
943     391           2
864     685           4
750     323           3
279     64            1
646     750           3
654     370           2
617     582           4
913     690           3
660     229           2
421     498           4
495     1091          4
806     421           4
676     538           4
721     262           3
913     209           2
378     78            3
880     476           3
716     204           5
276     1090          1
13      225           2
12      203           3

[100000 rows x 1 columns]

#### 인덱스컬럼으로 언스택

In [99]:
data.set_index(['user_id', 'item_id']).unstack()

rating                                               ...            \
item_id   1    2    3    4    5    6    7    8    9    10    ... 1673 1674   
user_id                                                      ...             
1          5.0  3.0  4.0  3.0  3.0  5.0  4.0  1.0  5.0  3.0  ...  NaN  NaN   
2          4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  2.0  ...  NaN  NaN   
3          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
4          NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
5          4.0  3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
6          4.0  NaN  NaN  NaN  NaN  NaN  2.0  4.0  4.0  NaN  ...  NaN  NaN   
7          NaN  NaN  NaN  5.0  NaN  NaN  5.0  5.0  5.0  4.0  ...  NaN  NaN   
8          NaN  NaN  NaN  NaN  NaN  NaN  3.0  NaN  NaN  NaN  ...  NaN  NaN   
9          NaN  NaN  NaN  NaN  NaN  5.0  4.0  NaN  NaN  NaN  ...  NaN  NaN   
10         4.0  NaN  NaN  4.0  NaN  NaN  4.0  NaN  4.0  NaN  ...  NaN  NaN   
11         NaN  NaN  NaN  NaN  NaN  NaN  NaN  4.0  5.0  NaN  ...  NaN  NaN   
12         NaN  NaN  NaN  5.0  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
13         3.0  3.0  NaN  5.0  1.0  NaN  2.0  4.0  3.0  NaN  ...  NaN  NaN   
14         NaN  NaN  NaN  NaN  NaN  NaN  5.0  NaN  4.0  NaN  ...  NaN  NaN   
15         1.0  NaN  NaN  NaN  NaN  NaN  1.0  NaN  4.0  NaN  ...  NaN  NaN   
16         5.0  NaN  NaN  5.0  NaN  NaN  5.0  5.0  5.0  NaN  ...  NaN  NaN   
17         4.0  NaN  NaN  NaN  NaN  NaN  4.0  NaN  3.0  NaN  ...  NaN  NaN   
18         5.0  NaN  NaN  3.0  NaN  5.0  NaN  5.0  5.0  NaN  ...  NaN  NaN   
19         NaN  NaN  NaN  4.0  NaN  NaN  NaN  5.0  NaN  NaN  ...  NaN  NaN   
20         3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
21         5.0  NaN  NaN  NaN  2.0  NaN  5.0  NaN  5.0  NaN  ...  NaN  NaN   
22         NaN  2.0  NaN  5.0  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
23         5.0  NaN  NaN  NaN  NaN  NaN  4.0  4.0  NaN  NaN  ...  NaN  NaN   
24         NaN  NaN  NaN  NaN  NaN  NaN  4.0  5.0  5.0  NaN  ...  NaN  NaN   
25         5.0  NaN  NaN  NaN  NaN  NaN  4.0  4.0  NaN  NaN  ...  NaN  NaN   
26         3.0  NaN  NaN  NaN  NaN  NaN  3.0  NaN  4.0  NaN  ...  NaN  NaN   
27         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  4.0  NaN  ...  NaN  NaN   
28         NaN  NaN  NaN  NaN  3.0  NaN  5.0  NaN  NaN  NaN  ...  NaN  NaN   
29         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
30         NaN  3.0  NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN  ...  NaN  NaN   
...        ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
914        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
915        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
916        4.0  3.0  3.0  4.0  3.0  NaN  4.0  NaN  5.0  NaN  ...  NaN  NaN   
917        3.0  NaN  1.0  NaN  NaN  NaN  NaN  NaN  5.0  NaN  ...  NaN  NaN   
918        3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
919        4.0  NaN  NaN  1.0  4.0  NaN  3.0  NaN  5.0  NaN  ...  NaN  NaN   
920        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
921        3.0  NaN  NaN  NaN  NaN  NaN  NaN  3.0  NaN  NaN  ...  NaN  NaN   
922        5.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
923        3.0  NaN  4.0  NaN  NaN  NaN  NaN  NaN  4.0  NaN  ...  NaN  NaN   
924        5.0  3.0  NaN  NaN  NaN  4.0  4.0  NaN  4.0  NaN  ...  NaN  NaN   
925        NaN  NaN  NaN  NaN  4.0  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
926        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
927        5.0  NaN  NaN  NaN  NaN  NaN  3.0  4.0  NaN  NaN  ...  NaN  NaN   
928        NaN  NaN  NaN  NaN  NaN  NaN  NaN  5.0  5.0  NaN  ...  NaN  NaN   
929        3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN   
930        3.0  NaN  NaN  NaN  NaN  NaN  NaN  3.0  NaN  NaN  ...  NaN  NaN   
931        NaN  NaN  NaN  NaN 

#### 이제 NaN을 0으로 바꿔야지
- fillna
- 

In [100]:
data.set_index(['user_id', 'item_id']).unstack().fillna(0)

rating                                               ...            \
item_id   1    2    3    4    5    6    7    8    9    10    ... 1673 1674   
user_id                                                      ...             
1          5.0  3.0  4.0  3.0  3.0  5.0  4.0  1.0  5.0  3.0  ...  0.0  0.0   
2          4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  2.0  ...  0.0  0.0   
3          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
4          0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
5          4.0  3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
6          4.0  0.0  0.0  0.0  0.0  0.0  2.0  4.0  4.0  0.0  ...  0.0  0.0   
7          0.0  0.0  0.0  5.0  0.0  0.0  5.0  5.0  5.0  4.0  ...  0.0  0.0   
8          0.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  ...  0.0  0.0   
9          0.0  0.0  0.0  0.0  0.0  5.0  4.0  0.0  0.0  0.0  ...  0.0  0.0   
10         4.0  0.0  0.0  4.0  0.0  0.0  4.0  0.0  4.0  0.0  ...  0.0  0.0   
11         0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0  5.0  0.0  ...  0.0  0.0   
12         0.0  0.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
13         3.0  3.0  0.0  5.0  1.0  0.0  2.0  4.0  3.0  0.0  ...  0.0  0.0   
14         0.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  4.0  0.0  ...  0.0  0.0   
15         1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  4.0  0.0  ...  0.0  0.0   
16         5.0  0.0  0.0  5.0  0.0  0.0  5.0  5.0  5.0  0.0  ...  0.0  0.0   
17         4.0  0.0  0.0  0.0  0.0  0.0  4.0  0.0  3.0  0.0  ...  0.0  0.0   
18         5.0  0.0  0.0  3.0  0.0  5.0  0.0  5.0  5.0  0.0  ...  0.0  0.0   
19         0.0  0.0  0.0  4.0  0.0  0.0  0.0  5.0  0.0  0.0  ...  0.0  0.0   
20         3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
21         5.0  0.0  0.0  0.0  2.0  0.0  5.0  0.0  5.0  0.0  ...  0.0  0.0   
22         0.0  2.0  0.0  5.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
23         5.0  0.0  0.0  0.0  0.0  0.0  4.0  4.0  0.0  0.0  ...  0.0  0.0   
24         0.0  0.0  0.0  0.0  0.0  0.0  4.0  5.0  5.0  0.0  ...  0.0  0.0   
25         5.0  0.0  0.0  0.0  0.0  0.0  4.0  4.0  0.0  0.0  ...  0.0  0.0   
26         3.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0  4.0  0.0  ...  0.0  0.0   
27         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  4.0  0.0  ...  0.0  0.0   
28         0.0  0.0  0.0  0.0  3.0  0.0  5.0  0.0  0.0  0.0  ...  0.0  0.0   
29         0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
30         0.0  3.0  0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0  ...  0.0  0.0   
...        ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
914        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
915        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
916        4.0  3.0  3.0  4.0  3.0  0.0  4.0  0.0  5.0  0.0  ...  0.0  0.0   
917        3.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  ...  0.0  0.0   
918        3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
919        4.0  0.0  0.0  1.0  4.0  0.0  3.0  0.0  5.0  0.0  ...  0.0  0.0   
920        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
921        3.0  0.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0  0.0  ...  0.0  0.0   
922        5.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
923        3.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  4.0  0.0  ...  0.0  0.0   
924        5.0  3.0  0.0  0.0  0.0  4.0  4.0  0.0  4.0  0.0  ...  0.0  0.0   
925        0.0  0.0  0.0  0.0  4.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
926        0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
927        5.0  0.0  0.0  0.0  0.0  0.0  3.0  4.0  0.0  0.0  ...  0.0  0.0   
928        0.0  0.0  0.0  0.0  0.0  0.0  0.0  5.0  5.0  0.0  ...  0.0  0.0   
929        3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
930        3.0  0.0  0.0  0.0  0.0  0.0  0.0  3.0  0.0  0.0  ...  0.0  0.0   
931        0.0  0.0  0.0  0.0 

#### 이제 우리가 만든거 아름다워. 자주 쓸수 있으니 이름 붙여

In [103]:
recommendation=data.set_index(['user_id', 'item_id']).unstack().fillna(0)

#### 원래 데이터하고 크로스탭하고 피봇테이블로 만들어봐
- pd.crosstab(data['광역시도'],data['년'], data['수'], aggfunc='count')
- data.pivot_table('수','년', aggfunc='count',margins=True)

### 크로스탭 내 시도

In [105]:
data=data_raw

In [122]:
data.drop(columns='timestamp', inplace=True)

In [123]:
data

user_id  item_id  rating
0          196      242       3
1          186      302       3
2           22      377       1
3          244       51       2
4          166      346       1
5          298      474       4
6          115      265       2
7          253      465       5
8          305      451       3
9            6       86       3
10          62      257       2
11         286     1014       5
12         200      222       5
13         210       40       3
14         224       29       3
15         303      785       3
16         122      387       5
17         194      274       2
18         291     1042       4
19         234     1184       2
20         119      392       4
21         167      486       4
22         299      144       4
23         291      118       2
24         308        1       4
25          95      546       2
26          38       95       5
27         102      768       2
28          63      277       4
29         160      234       5
...        ...      ...     ...
99970      449      120       1
99971      661      762       2
99972      721      874       3
99973      821      151       4
99974      764      596       3
99975      537      443       3
99976      618      628       2
99977      487      291       3
99978      113      975       5
99979      943      391       2
99980      864      685       4
99981      750      323       3
99982      279       64       1
99983      646      750       3
99984      654      370       2
99985      617      582       4
99986      913      690       3
99987      660      229       2
99988      421      498       4
99989      495     1091       4
99990      806      421       4
99991      676      538       4
99992      721      262       3
99993      913      209       2
99994      378       78       3
99995      880      476       3
99996      716      204       5
99997      276     1090       1
99998       13      225       2
99999       12      203       3

[100000 rows x 3 columns]

### 집계함수 삼총사
- 그룹바이, 크로스탭, 피봇

## Pivot

In [130]:
data.pivot_table('rating',index=["user_id"], columns=["item_id"], aggfunc="sum").fillna(0)

item_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         4.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   2.0  ...   
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
5         4.0   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
6         4.0   NaN   NaN   NaN   NaN   NaN   2.0   4.0   4.0   NaN  ...   
7         NaN   NaN   NaN   5.0   NaN   NaN   5.0   5.0   5.0   4.0  ...   
8         NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN   NaN  ...   
9         NaN   NaN   NaN   NaN   NaN   5.0   4.0   NaN   NaN   NaN  ...   
10        4.0   NaN   NaN   4.0   NaN   NaN   4.0   NaN   4.0   NaN  ...   
11        NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0   5.0   NaN  ...   
12        NaN   NaN   NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN  ...   
13        3.0   3.0   NaN   5.0   1.0   NaN   2.0   4.0   3.0   NaN  ...   
14        NaN   NaN   NaN   NaN   NaN   NaN   5.0   NaN   4.0   NaN  ...   
15        1.0   NaN   NaN   NaN   NaN   NaN   1.0   NaN   4.0   NaN  ...   
16        5.0   NaN   NaN   5.0   NaN   NaN   5.0   5.0   5.0   NaN  ...   
17        4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   3.0   NaN  ...   
18        5.0   NaN   NaN   3.0   NaN   5.0   NaN   5.0   5.0   NaN  ...   
19        NaN   NaN   NaN   4.0   NaN   NaN   NaN   5.0   NaN   NaN  ...   
20        3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
21        5.0   NaN   NaN   NaN   2.0   NaN   5.0   NaN   5.0   NaN  ...   
22        NaN   2.0   NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN  ...   
23        5.0   NaN   NaN   NaN   NaN   NaN   4.0   4.0   NaN   NaN  ...   
24        NaN   NaN   NaN   NaN   NaN   NaN   4.0   5.0   5.0   NaN  ...   
25        5.0   NaN   NaN   NaN   NaN   NaN   4.0   4.0   NaN   NaN  ...   
26        3.0   NaN   NaN   NaN   NaN   NaN   3.0   NaN   4.0   NaN  ...   
27        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0   NaN  ...   
28        NaN   NaN   NaN   NaN   3.0   NaN   5.0   NaN   NaN   NaN  ...   
29        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
30        NaN   3.0   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
914       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
915       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
916       4.0   3.0   3.0   4.0   3.0   NaN   4.0   NaN   5.0   NaN  ...   
917       3.0   NaN   1.0   NaN   NaN   NaN   NaN   NaN   5.0   NaN  ...   
918       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
919       4.0   NaN   NaN   1.0   4.0   NaN   3.0   NaN   5.0   NaN  ...   
920       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
921       3.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...   
922       5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
923       3.0   NaN   4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN  ...   
924       5.0   3.0   NaN   NaN   NaN   4.0   4.0   NaN   4.0   NaN  ...   
925       NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN   NaN   NaN  ...   
926       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
927       5.0   NaN   NaN   NaN   NaN   NaN   3.0   4.0   NaN   NaN  ...   
928       NaN   NaN   NaN   NaN   NaN   NaN   NaN   5.0   5.0   NaN  ...   
929       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
930       3.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...   
931       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
932       4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   5.0   NaN  ...   
933       3.0   NaN   NaN   3.0   NaN   NaN   4.

In [135]:
data.pivot('user_id','item_id','rating')

item_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         4.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   2.0  ...   
3         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
4         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
5         4.0   3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
6         4.0   NaN   NaN   NaN   NaN   NaN   2.0   4.0   4.0   NaN  ...   
7         NaN   NaN   NaN   5.0   NaN   NaN   5.0   5.0   5.0   4.0  ...   
8         NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN   NaN  ...   
9         NaN   NaN   NaN   NaN   NaN   5.0   4.0   NaN   NaN   NaN  ...   
10        4.0   NaN   NaN   4.0   NaN   NaN   4.0   NaN   4.0   NaN  ...   
11        NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0   5.0   NaN  ...   
12        NaN   NaN   NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN  ...   
13        3.0   3.0   NaN   5.0   1.0   NaN   2.0   4.0   3.0   NaN  ...   
14        NaN   NaN   NaN   NaN   NaN   NaN   5.0   NaN   4.0   NaN  ...   
15        1.0   NaN   NaN   NaN   NaN   NaN   1.0   NaN   4.0   NaN  ...   
16        5.0   NaN   NaN   5.0   NaN   NaN   5.0   5.0   5.0   NaN  ...   
17        4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   3.0   NaN  ...   
18        5.0   NaN   NaN   3.0   NaN   5.0   NaN   5.0   5.0   NaN  ...   
19        NaN   NaN   NaN   4.0   NaN   NaN   NaN   5.0   NaN   NaN  ...   
20        3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
21        5.0   NaN   NaN   NaN   2.0   NaN   5.0   NaN   5.0   NaN  ...   
22        NaN   2.0   NaN   5.0   NaN   NaN   NaN   NaN   NaN   NaN  ...   
23        5.0   NaN   NaN   NaN   NaN   NaN   4.0   4.0   NaN   NaN  ...   
24        NaN   NaN   NaN   NaN   NaN   NaN   4.0   5.0   5.0   NaN  ...   
25        5.0   NaN   NaN   NaN   NaN   NaN   4.0   4.0   NaN   NaN  ...   
26        3.0   NaN   NaN   NaN   NaN   NaN   3.0   NaN   4.0   NaN  ...   
27        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   4.0   NaN  ...   
28        NaN   NaN   NaN   NaN   3.0   NaN   5.0   NaN   NaN   NaN  ...   
29        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
30        NaN   3.0   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
914       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
915       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
916       4.0   3.0   3.0   4.0   3.0   NaN   4.0   NaN   5.0   NaN  ...   
917       3.0   NaN   1.0   NaN   NaN   NaN   NaN   NaN   5.0   NaN  ...   
918       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
919       4.0   NaN   NaN   1.0   4.0   NaN   3.0   NaN   5.0   NaN  ...   
920       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
921       3.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...   
922       5.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
923       3.0   NaN   4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN  ...   
924       5.0   3.0   NaN   NaN   NaN   4.0   4.0   NaN   4.0   NaN  ...   
925       NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN   NaN   NaN  ...   
926       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
927       5.0   NaN   NaN   NaN   NaN   NaN   3.0   4.0   NaN   NaN  ...   
928       NaN   NaN   NaN   NaN   NaN   NaN   NaN   5.0   5.0   NaN  ...   
929       3.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
930       3.0   NaN   NaN   NaN   NaN   NaN   NaN   3.0   NaN   NaN  ...   
931       NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
932       4.0   NaN   NaN   NaN   NaN   NaN   4.0   NaN   5.0   NaN  ...   
933       3.0   NaN   NaN   3.0   NaN   NaN   4.

In [136]:
reco=data.pivot('user_id','item_id','rating').fillna(0)

## pd.crosstab
- 설명 보면 얘는 aggfunc None이라 맥스 넣든 민 넣듯 무관

In [132]:
pd.crosstab(data.user_id, data.item_id,data.rating, aggfunc='max').fillna(0)

item_id  1     2     3     4     5     6     7     8     9     10    ...  \
user_id                                                              ...   
1         5.0   3.0   4.0   3.0   3.0   5.0   4.0   1.0   5.0   3.0  ...   
2         4.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   2.0  ...   
3         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
4         0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
5         4.0   3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
6         4.0   0.0   0.0   0.0   0.0   0.0   2.0   4.0   4.0   0.0  ...   
7         0.0   0.0   0.0   5.0   0.0   0.0   5.0   5.0   5.0   4.0  ...   
8         0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0   0.0  ...   
9         0.0   0.0   0.0   0.0   0.0   5.0   4.0   0.0   0.0   0.0  ...   
10        4.0   0.0   0.0   4.0   0.0   0.0   4.0   0.0   4.0   0.0  ...   
11        0.0   0.0   0.0   0.0   0.0   0.0   0.0   4.0   5.0   0.0  ...   
12        0.0   0.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
13        3.0   3.0   0.0   5.0   1.0   0.0   2.0   4.0   3.0   0.0  ...   
14        0.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0   4.0   0.0  ...   
15        1.0   0.0   0.0   0.0   0.0   0.0   1.0   0.0   4.0   0.0  ...   
16        5.0   0.0   0.0   5.0   0.0   0.0   5.0   5.0   5.0   0.0  ...   
17        4.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0   3.0   0.0  ...   
18        5.0   0.0   0.0   3.0   0.0   5.0   0.0   5.0   5.0   0.0  ...   
19        0.0   0.0   0.0   4.0   0.0   0.0   0.0   5.0   0.0   0.0  ...   
20        3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
21        5.0   0.0   0.0   0.0   2.0   0.0   5.0   0.0   5.0   0.0  ...   
22        0.0   2.0   0.0   5.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
23        5.0   0.0   0.0   0.0   0.0   0.0   4.0   4.0   0.0   0.0  ...   
24        0.0   0.0   0.0   0.0   0.0   0.0   4.0   5.0   5.0   0.0  ...   
25        5.0   0.0   0.0   0.0   0.0   0.0   4.0   4.0   0.0   0.0  ...   
26        3.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   4.0   0.0  ...   
27        0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0  ...   
28        0.0   0.0   0.0   0.0   3.0   0.0   5.0   0.0   0.0   0.0  ...   
29        0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
30        0.0   3.0   0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0  ...   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   
914       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
915       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
916       4.0   3.0   3.0   4.0   3.0   0.0   4.0   0.0   5.0   0.0  ...   
917       3.0   0.0   1.0   0.0   0.0   0.0   0.0   0.0   5.0   0.0  ...   
918       3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
919       4.0   0.0   0.0   1.0   4.0   0.0   3.0   0.0   5.0   0.0  ...   
920       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
921       3.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0  ...   
922       5.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
923       3.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0  ...   
924       5.0   3.0   0.0   0.0   0.0   4.0   4.0   0.0   4.0   0.0  ...   
925       0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0  ...   
926       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
927       5.0   0.0   0.0   0.0   0.0   0.0   3.0   4.0   0.0   0.0  ...   
928       0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0   5.0   0.0  ...   
929       3.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
930       3.0   0.0   0.0   0.0   0.0   0.0   0.0   3.0   0.0   0.0  ...   
931       0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  ...   
932       4.0   0.0   0.0   0.0   0.0   0.0   4.0   0.0   5.0   0.0  ...   
933       3.0   0.0   0.0   3.0   0.0   0.0   4.

In [41]:
#u.occupation
occupation=pd.read_csv('C:\Python\ml-100k\ml-100k/u.occupation', header=None, )

In [42]:
occupation

0
0   administrator
1          artist
2          doctor
3        educator
4        engineer
5   entertainment
6       executive
7      healthcare
8       homemaker
9          lawyer
10      librarian
11      marketing
12           none
13          other
14     programmer
15        retired
16       salesman
17      scientist
18        student
19     technician
20         writer

In [ ]:
item

In [86]:
#u.user불러오기
user= pd.read_csv('C:/Python/ml-100k/ml-100k/u.user', header=None, engine='python' ,sep='|')

In [87]:
user.melt

<bound method DataFrame.melt of        0   1  2              3      4
0      1  24  M     technician  85711
1      2  53  F          other  94043
2      3  23  M         writer  32067
3      4  24  M     technician  43537
4      5  33  F          other  15213
5      6  42  M      executive  98101
6      7  57  M  administrator  91344
7      8  36  M  administrator  05201
8      9  29  M        student  01002
9     10  53  M         lawyer  90703
10    11  39  F          other  30329
11    12  28  F          other  06405
12    13  47  M       educator  29206
13    14  45  M      scientist  55106
14    15  49  F       educator  97301
15    16  21  M  entertainment  10309
16    17  30  M     programmer  06355
17    18  35  F          other  37212
18    19  40  M      librarian  02138
19    20  42  F      homemaker  95660
20    21  26  M         writer  30068
21    22  25  M         writer  40206
22    23  30  F         artist  48197
23    24  21  F         artist  94533
24    25  39  M   

In [88]:
user.rename({0:'user_id', 1:'age',2:'gender', 3:'occupation', 4:'zip_code'}, axis=1, inplace=True)

## 이제 피봇/크로스탭 등으로 다듬었으니, 유사도 해야해요

### 코릴레이션
- 어제 뭐해서 나왔어요? 
- 컬럼끼리인데 시리즈 아니고 데이터 테이블 만들기 위해서
    - 한개짜리 만들려면? 팬시 인덱싱

In [137]:
data.corr()

user_id   item_id    rating
user_id  1.000000  0.010377 -0.009371
item_id  0.010377  1.000000 -0.189124
rating  -0.009371 -0.189124  1.000000

In [140]:
data[['user_id']].corr()

user_id
user_id      1.0

In [141]:
reco.corr()

item_id      1         2         3         4         5         6         7     \
item_id                                                                         
1        1.000000  0.234595  0.193362  0.226213  0.128840  0.015113  0.347354   
2        0.234595  1.000000  0.190649  0.409044  0.240712  0.030062  0.220022   
3        0.193362  0.190649  1.000000  0.227849  0.141368  0.065347  0.258855   
4        0.226213  0.409044  0.227849  1.000000  0.237298  0.021878  0.295489   
5        0.128840  0.240712  0.141368  0.237298  1.000000 -0.008594  0.205289   
6        0.015113  0.030062  0.065347  0.021878 -0.008594  1.000000  0.054415   
7        0.347354  0.220022  0.258855  0.295489  0.205289  0.054415  1.000000   
8        0.254490  0.206020  0.078636  0.352800  0.145866  0.012330  0.190670   
9        0.209502  0.077894  0.146181  0.229922  0.142541  0.079619  0.286572   
10       0.104655  0.072906  0.079608  0.138220 -0.033746  0.166084  0.178505   
11       0.221811  0.362190  0.265441  0.473171  0.281260 -0.013182  0.391409   
12       0.171780  0.343212  0.207190  0.457358  0.272113  0.074459  0.364064   
13       0.198063  0.074405  0.172782  0.254039  0.084938  0.228730  0.287245   
14       0.089739 -0.032029  0.078041  0.090185 -0.024039  0.209423  0.090469   
15       0.339027  0.076977  0.138165  0.113703  0.134558  0.010064  0.149448   
16       0.129048  0.031636  0.119143  0.092877  0.059505  0.115907  0.117292   
17       0.108860  0.320204  0.349635  0.319291  0.332488  0.045678  0.293315   
18      -0.023226  0.085478  0.104404  0.065539  0.052929  0.228790  0.028663   
19       0.022113 -0.069434  0.032821  0.012338  0.013776  0.186435  0.102364   
20       0.080876 -0.001471  0.122217  0.104686 -0.015551  0.152543  0.168835   
21       0.212938  0.244859  0.124356  0.151499  0.121435  0.030429  0.167264   
22       0.256862  0.368449  0.170528  0.434904  0.200503 -0.008163  0.272336   
23       0.072484  0.147752  0.127550  0.312218  0.139066  0.121458  0.188074   
24       0.281745  0.338009  0.301930  0.278431  0.134909  0.056764  0.365520   
25       0.333410  0.146921  0.189111  0.238576  0.159652  0.036667  0.310012   
26       0.064470  0.122634  0.148284  0.300564  0.163444  0.114321  0.111280   
27       0.072615  0.358680  0.208946  0.305548  0.188386  0.024980  0.192020   
28       0.376461  0.371013  0.144225  0.384661  0.213164 -0.000004  0.229629   
29       0.204595  0.526570  0.176465  0.335032  0.250817  0.018161  0.171396   
30       0.028232  0.022024  0.075252  0.147436  0.076392  0.162039  0.083848   
...           ...       ...       ...       ...       ...       ...       ...   
1653    -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1654     0.050142 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1655     0.034179  0.071415  0.091421  0.046743  0.088618 -0.005159  0.039920   
1656     0.005893 -0.016191 -0.012697  0.056858 -0.012859 -0.006708  0.042721   
1657     0.018215  0.099371 -0.009764 -0.016619  0.088618 -0.005159  0.006942   
1658     0.015635 -0.020829 -0.016334 -0.027801 -0.016543 -0.008630  0.033067   
1659     0.050142 -0.012451  0.057693  0.067863 -0.009889 -0.005159  0.039920   
1660     0.050142 -0.012451  0.057693  0.067863 -0.009889 -0.005159  0.039920   
1661     0.018215  0.099371  0.091421 -0.016619 -0.009889 -0.005159  0.023431   
1662    -0.035996 -0.015103 -0.011844 -0.020158 -0.011995 -0.006258 -0.031582   
1663    -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1664     0.036156  0.056823  0.001141  0.041604  0.037974  0.110130  0.022343   
1665    -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1666    -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1667    -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1668    -0.029676 -0.012451 -0.009764 -0.016619 -0.009889 -0.005159 -0.026036   
1669    -0.029676 -0.012451 -0.

In [6]:
reco.T

NameError: name 'reco' is not defined

In [143]:
reco.T.corr()

user_id       1         2         3         4         5         6         7    \
user_id                                                                         
1        1.000000  0.106322 -0.016424  0.021716  0.302592  0.345976  0.320017   
2        0.106322  1.000000  0.082680  0.160323  0.020218  0.197844  0.022886   
3       -0.016424  0.082680  1.000000  0.331674 -0.028322  0.017673 -0.013384   
4        0.021716  0.160323  0.331674  1.000000 -0.002426  0.030206  0.041056   
5        0.302592  0.020218 -0.028322 -0.002426  1.000000  0.153409  0.279677   
6        0.345976  0.197844  0.017673  0.030206  0.153409  1.000000  0.396173   
7        0.320017  0.022886 -0.013384  0.041056  0.279677  0.396173  1.000000   
8        0.276208  0.072772  0.055385  0.171116  0.209280  0.152073  0.231497   
9        0.038812  0.143716  0.043505  0.089372  0.025496  0.156114  0.105950   
10       0.288491  0.106861  0.012080  0.024018  0.117166  0.495458  0.398928   
11       0.217507  0.075766  0.025442  0.071262  0.252772  0.171418  0.269335   
12       0.273815  0.092465  0.040268  0.056481  0.085580  0.195640  0.217795   
13       0.296534  0.147474  0.129876  0.066668  0.229683  0.361222  0.404698   
14       0.270445  0.191993  0.004462  0.044836  0.189778  0.356823  0.251591   
15       0.106351  0.432019  0.096585  0.126110  0.032820  0.128333  0.009152   
16       0.335007  0.077818  0.013078  0.054693  0.224095  0.329387  0.358805   
17       0.184958  0.206313  0.011647  0.040303  0.052050  0.174693  0.044772   
18       0.380732  0.117309 -0.034706 -0.011837  0.198013  0.521402  0.410341   
19       0.074695  0.083600  0.088497  0.148089  0.090244  0.100071  0.116162   
20       0.198485  0.032104  0.025274  0.069366  0.205944  0.137041  0.161568   
21       0.098842  0.159973  0.124820  0.163750  0.099764  0.043699  0.137534   
22       0.318704  0.018688 -0.007348  0.054926  0.412184  0.175814  0.291707   
23       0.379766  0.098722 -0.017713  0.029479  0.354731  0.382413  0.329415   
24       0.231979  0.147140  0.040256  0.156495  0.085329  0.271222  0.238793   
25       0.281809  0.113763  0.002089  0.061020  0.286415  0.334610  0.277153   
26       0.178358  0.420319  0.129180  0.127635  0.087925  0.150112  0.027637   
27       0.092360  0.167306  0.001982  0.061360  0.048942  0.094328  0.052979   
28       0.246582  0.066191  0.034740  0.130522  0.279251  0.151972  0.277039   
29       0.083585  0.153623  0.133598  0.129412  0.012533  0.139959  0.081523   
30       0.134798  0.207484  0.085206  0.128847  0.126665  0.126022  0.092844   
...           ...       ...       ...       ...       ...       ...       ...   
914      0.017865  0.019657 -0.018186 -0.012844  0.007465  0.005642  0.017822   
915      0.002821  0.244619  0.345865  0.124093 -0.036568  0.043603  0.010074   
916      0.487949  0.043629 -0.025589  0.006625  0.253080  0.303088  0.366664   
917      0.097007  0.266140  0.022542  0.028358  0.036331  0.093542  0.041109   
918      0.159046  0.006593 -0.033843 -0.028503  0.107855  0.218266  0.190244   
919      0.238376  0.339394  0.224201  0.152497  0.094282  0.209205  0.113308   
920      0.014200  0.314020  0.423346  0.159370 -0.037063  0.035294 -0.001475   
921      0.206190  0.128491  0.057450  0.074810  0.202748  0.139026  0.156350   
922      0.317505  0.077631  0.012516  0.106218  0.375348  0.137561  0.295133   
923      0.088004  0.201167  0.134918  0.072190  0.105465  0.070048  0.044477   
924      0.244603  0.213287  0.035108  0.058690  0.158590  0.284077  0.291808   
925      0.031617  0.024560  0.166592  0.155011  0.026047  0.013180  0.061961   
926      0.047844  0.379684  0.274150  0.222635 -0.032250  0.101744  0.023803   
927      0.179413  0.070114  0.011324  0.101712  0.232972  0.065795  0.176718   
928      0.199129  0.067676  0.048688  0.038013  0.047889  0.201839  0.110749   
929      0.238361  0.086790  0.009863  0.029559  0.196578  0.402987  0.303467   
930      0.231222  0.263913 -0.

### 이제 특별한 사람 정해요
- 여기서 42번  뽑아두어요

## 할당

In [144]:
reco_corr=reco.T.corr()

In [146]:
reco_corr.loc[42]

user_id
1      0.288139
2      0.061978
3     -0.012050
4      0.035344
5      0.281515
6      0.226767
7      0.292792
8      0.210809
9      0.094708
10     0.188192
11     0.264883
12     0.349561
13     0.287686
14     0.182963
15     0.140365
16     0.400000
17     0.066929
18     0.284390
19     0.068612
20     0.319723
21     0.016978
22     0.262085
23     0.338755
24     0.183717
25     0.288307
26     0.127056
27     0.052150
28     0.234852
29     0.020179
30     0.114274
         ...   
914    0.137118
915   -0.040124
916    0.341528
917    0.089095
918    0.126220
919    0.179692
920   -0.040668
921    0.405436
922    0.387550
923    0.169071
924    0.265752
925    0.000500
926   -0.003651
927    0.362487
928    0.075156
929    0.196203
930    0.152169
931    0.041755
932    0.190091
933    0.404733
934    0.257091
935    0.204358
936    0.033791
937    0.012802
938    0.146442
939    0.178775
940    0.187521
941    0.065588
942    0.203358
943    0.367843
Name: 42, Length

### 이건 i 붙이나 안 붙이나 똑같죠. 
이제 소트

In [147]:
reco_corr.loc[42].sort_values(ascending=False)

user_id
42     1.000000
577    0.513391
290    0.502035
864    0.501236
311    0.496347
484    0.494650
618    0.492603
882    0.490165
897    0.482251
632    0.478189
796    0.470999
330    0.470584
749    0.469258
533    0.461316
786    0.461007
815    0.460907
378    0.459101
43     0.457879
487    0.457623
222    0.457292
899    0.451498
301    0.446393
230    0.444312
606    0.443007
554    0.441804
892    0.439502
56     0.439420
200    0.438084
716    0.437350
128    0.435804
         ...   
713   -0.043204
898   -0.043230
845   -0.043328
720   -0.044274
841   -0.044299
112   -0.044480
427   -0.045205
149   -0.046167
448   -0.046822
100   -0.047147
206   -0.047158
695   -0.047242
179   -0.047441
828   -0.047633
428   -0.049206
787   -0.049502
784   -0.050772
611   -0.051442
724   -0.051506
616   -0.052944
574   -0.053133
351   -0.053473
519   -0.054804
284   -0.055311
752   -0.058097
451   -0.060422
863   -0.069825
587   -0.070266
489   -0.078078
782   -0.106432
Name: 42, Length

In [148]:
reco_corr.loc[42].sort_values(ascending=False).iloc[1:5] # 내꺼 빼야지. iloc 왜 쓰니?차례로 안 되어있으니 차례대로 하면 아이엘오씨하면 나 빼고 1에서 오까지

user_id
577    0.513391
290    0.502035
864    0.501236
311    0.496347
Name: 42, dtype: float64

- 내꺼 빼야지. iloc 왜 쓰니?차례로 안 되어있으니 차례대로 하면 아이엘오씨하면 나 빼고 1에서 오까지
### 위의 결과는 우아하지 않아. 함수 만들어

## 재활용 리코멘데이션

In [150]:
def recommendation_user(user_id, k):
    return reco_corr.loc[user_id].sort_values(ascending=False)[1:k]

In [151]:
recommendation_user(1,5)

user_id
916    0.487949
738    0.471105
864    0.465283
457    0.457111
Name: 1, dtype: float64

In [174]:
s=recommendation_user(1,5)

In [175]:
s

user_id
916    0.487949
738    0.471105
864    0.465283
457    0.457111
Name: 1, dtype: float64

In [176]:
for i in s:
    print(i) #포를 돌릴때 

0.4879491330327622
0.4711048400073322
0.4652827485719093
0.457111312198188


In [177]:
s.index

Int64Index([916, 738, 864, 457], dtype='int64', name='user_id')

#### 이렇게 인덱스만 뽑고 싶다면
- 916, 738, 864, 457],
- def recommendation_user(user_id, k):
    return reco_corr.loc[user_id].sort_values(ascending=False)[1:k] 
    - 여기에서
- def recommendation_user(user_id, k):
    return reco_corr.loc[user_id].sort_values(ascending=False)[1:k].index


#### concat
- s와 s1 테이블 붙이기(axis=0)

In [ ]:
pd.concat([s,s1])

In [ ]:
#### duplicated
- 중복 여부
- 중복 지우려면 drop duplicated

In [ ]:
x=pd.concat([s,s1]).duplicated()

In [ ]:
x.drop_duplicates()

In [ ]:
#### 남기는거 없애 
- keep=False

In [ ]:
x.drop_duplicates(keep=False)

In [188]:
# 데이터프레임 만들기
temp=pd.DataFrame(['user_id', 'item_id','ranking'])

In [189]:
temp

0
0  user_id
1  item_id
2  ranking

In [4]:
def recommendation_user(user_id, k):
    s=reco_corr.loc[user_id].sort_values(ascending=False)[1:k].index
    temp=pd.DataFrame(columns=['user_id', 'item_id','ranting'])
    for i in s:
        t=set(data[data.user_id==i].item_id)- set(data[data.user_id==user_id].item_id)
        temp=pd.concat([temp,data[data.item_id.isin(t)& (data.rating==5) & (data.user_id==i)])
    else:
        return temp.drop_dyokucated(keep=False).sort_values('item_id')

SyntaxError: invalid syntax (<ipython-input-4-0bdd2632d015>, line 6)

In [5]:
aaa = recommendation_user(42,5,3)

NameError: name 'reco_corr' is not defined

In [190]:
s

## 이제 내가 본 영화라 쟤가 본 영화 걸러야해
- 데이터에서 유저아이디가 42인애 묶어내

In [154]:
data[data.user_id==42].item_id

58        423
87        403
111        96
166       794
310       546
638       274
659       588
754        44
778      1028
871       625
1119       98
1245      953
1656      685
1761      176
1864      195
2179      185
2290       69
2394      684
2431      603
2613     1041
2809      121
2833      606
2863      456
2979      506
3060      143
3162      746
3480       48
3542      999
3601      102
3824      210
         ... 
59045    1046
60647      97
61040     559
61695      86
62662      71
63521     941
67740     655
68778     969
69604     132
70257     595
70653    1048
71380     826
73843     111
74673    1050
76940     141
77201     181
79562     174
79923      73
82241     168
84817       2
85776     658
86538     755
86768     282
87394     136
88397     735
89527     411
93817     433
94791      87
97557     591
98433     103
Name: item_id, Length: 183, dtype: int64

### 아름답게 소트

In [155]:
data[data.user_id==42].item_id.sort_values()

20548       1
84817       2
8184       12
21019      15
6844       25
8789       28
47186      38
25807      43
754        44
3480       48
55954      50
25001      54
27013      58
16939      63
44706      64
36993      66
2290       69
4613       70
62662      71
17744      72
79923      73
18663      77
12099      79
8353       82
22743      83
61695      86
94791      87
8709       88
17887      95
111        96
         ... 
4297      781
3835      785
166       794
71380     826
12967     834
4277      845
19422     866
31466     924
7362      925
47516     926
26896     934
10320     939
63521     941
1245      953
68778     969
15895     977
3542      999
778      1028
11085    1040
2613     1041
21794    1042
5009     1043
10716    1044
12708    1045
59045    1046
20791    1047
70653    1048
17820    1049
74673    1050
7120     1051
Name: item_id, Length: 183, dtype: int64

In [158]:
set(data[data.user_id==577].item_id)- set(data[data.user_id==42].item_id)

{3,
 4,
 5,
 7,
 9,
 11,
 14,
 17,
 22,
 23,
 24,
 30,
 31,
 33,
 39,
 42,
 46,
 49,
 51,
 52,
 53,
 55,
 56,
 60,
 65,
 68,
 76,
 80,
 81,
 85,
 89,
 90,
 91,
 92,
 100,
 101,
 106,
 109,
 117,
 123,
 134,
 137,
 144,
 147,
 148,
 150,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 163,
 164,
 170,
 171,
 177,
 179,
 180,
 182,
 186,
 188,
 190,
 192,
 193,
 198,
 206,
 209,
 212,
 213,
 214,
 217,
 218,
 221,
 223,
 226,
 232,
 233,
 235,
 236,
 238,
 241,
 244,
 246,
 249,
 250,
 252,
 256,
 257,
 268,
 271,
 286,
 295,
 298,
 317,
 356,
 366,
 381,
 382,
 393,
 399,
 417,
 421,
 425,
 431,
 461,
 470,
 472,
 474,
 475,
 476,
 480,
 483,
 484,
 498,
 509,
 511,
 512,
 527,
 528,
 530,
 531,
 535,
 537,
 541,
 549,
 550,
 557,
 558,
 561,
 569,
 570,
 578,
 581,
 583,
 593,
 597,
 631,
 636,
 640,
 642,
 650,
 652,
 674,
 678,
 693,
 697,
 702,
 704,
 708,
 709,
 710,
 713,
 715,
 721,
 727,
 737,
 739,
 741,
 748,
 762,
 763,
 764,
 767,
 790,
 792,
 806,
 820,
 824,
 825,
 831,
 

In [161]:
t=set(data[data.user_id==577].item_id)- set(data[data.user_id==42].item_id)

## 이 순서를 어떻게 추천할까
- 우리 평점있으니 추천해주겠죠
- 이즈인태크닉써요

In [165]:
data[data.item_id.isin(t)].sort_values('rating',ascending=False)

user_id  item_id  rating
79685      340      480       5
92653      796      147       5
15227       21      100       5
15222       87      186       5
85068      406       56       5
24156      296      100       5
76530      457      154       5
76532      550      271       5
55265      757      549       5
15201      160      475       5
48180      394      186       5
55235      326      226       5
55269       38      144       5
42715      115      137       5
48179      707      286       5
45587      627      177       5
55283       85      483       5
92660      864       24       5
15153      276       91       5
55285      318      100       5
72853      370      650       5
55230      178      792       5
55424      561      652       5
31188       10      134       5
92617      299       89       5
55157      681      286       5
77925      737      192       5
72918      640       55       5
48211      577      549       5
92628      923      257       5
...        ...      ...     ...
32702      582      948       1
32753      405     1206       1
54507       91       56       1
44815      648       68       1
7436       125      475       1
44820      626      948       1
32831      486      250       1
54401      104      271       1
32832      263      271       1
54271      387      241       1
54713      682      476       1
73147      717      250       1
32464      524      541       1
32596      561      109       1
20329      363      557       1
84580      653      356       1
32475       44      737       1
7087       346      727       1
12897      405      721       1
32530      445      235       1
78266      296      244       1
7117       297      250       1
54953      774       31       1
7129       255      569       1
12884      255      748       1
32584      462      271       1
96424      774      561       1
12861       20      931       1
44785      684      218       1
83565      179      271       1

[24153 rows x 3 columns]

# 이 영화 추천 예비 후보

In [166]:
data[data.item_id.isin(t)& (data.rating==5)&(data.user_id==577)]

user_id  item_id  rating
36153      577      727       5
40266      577      241       5
47055      577      217       5
48211      577      549       5
48731      577       65       5
69872      577      470       5
76947      577      317       5
90653      577       22       5

In [168]:
r1=data[data.item_id.isin(t)& (data.rating==5)&(data.user_id==577)]

### 여기서 가벼운거 가중치 곱해줘요

In [169]:
r1.rating=r1.rating*0.513391

C:\Users\missr\Anaconda3\lib\site-packages\pandas\core\generic.py:5096: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [170]:
r1

user_id  item_id    rating
36153      577      727  2.566955
40266      577      241  2.566955
47055      577      217  2.566955
48211      577      549  2.566955
48731      577       65  2.566955
69872      577      470  2.566955
76947      577      317  2.566955
90653      577       22  2.566955

In [171]:
##

In [172]:
r1.rating=r1.rating*0.9

In [173]:
r1

user_id  item_id   rating
36153      577      727  2.31026
40266      577      241  2.31026
47055      577      217  2.31026
48211      577      549  2.31026
48731      577       65  2.31026
69872      577      470  2.31026
76947      577      317  2.31026
90653      577       22  2.31026

In [ ]:
temp=pd.DataFrame(colums=['user_id','item_id','ranking']) #템프 비어있다가 함수 진행되며 추가됨

In [2]:
def recommendation_user(user_id, k,n=10): #초기값 영 넣고
    s=reco_corr.loc[user_id].sort_values(ascending=False)[1:k+1].index #나하고 비슷한애 찾는ㄱ
    temp=pd.DataFrame(columns=['user_id', 'item_id','ranting'])
    for j,i in enumerate(s,1): #start 있음 enumurate는 인덱스 붙여 1,2,3,4돼
        t=set(data[data.user_id==i].item_id)- set(data[data.user_id==user_id].item_id)
        k=data[(data.item_id.isin(t)) & (data.rating==5) & (data.user_id==i)]
        k.rating=k.rating*(1-0.05*j)#밑에있을수록 005씩 가중치 줄여서 붙이기. j 0qnxj 
        pd.concat([temp,k])
        
    else:
        #temp= temp.rename(columns={'item_id':'movie_id'},inplace=True)#뮤터블이니까 빼고 
        #temp= temp.rename(columns={'item_id':'movie_id'})###@@@@@@@@@@@@머지붙이고, n개 추천하는거 코드 넣어 @@@@@
        temp=temp.rename(columns={'item_id':'movie_id'}, inplace=True) #이렇게 하면 뮤터블 NOne이라  안 나오니까 재할당 필요
        return temp

In [241]:
#aaa=recommendation_user(42,5,10)#@@@@@@@@@@@@@@@@@@@@@@@ 42번애가 가까운 5개 대해 조사하고, 10개 추천

In [3]:
aaa=recommendation_user(42,5)

NameError: name 'reco_corr' is not defined

In [243]:
aaa

In [ ]:
aaa.index

In [ ]:
[x for x in aaa]''

In [244]:
t=reco_corr.loc[42].sort_values(ascending=False)[1:5].index

In [245]:
aaa.merge(item) #머지하면 컬럼끼리 이너조인(관련된거 붙여줘)

AttributeError: 'NoneType' object has no attribute 'merge'

In [246]:
aaa.merge(item[['movie_id','movie_title']])

AttributeError: 'NoneType' object has no attribute 'merge'

In [239]:
aaa.merge(item[['movie_id','movie_title']]).drop_duplicates() # 기본 옵션이 킵 몇개 남길까. 

AttributeError: 'NoneType' object has no attribute 'merge'

In [247]:
aaa.merge(item[['movie_id','movie_title']]).drop_duplicates().groupby('movie_id').mean() # 평균....?

AttributeError: 'NoneType' object has no attribute 'merge'

In [209]:
for i in t:
    print(i)

577
290
864
311


# 여기서 문제 발생. ?@@@
- 드랍 듀플리케이트는 모든게 동일해야 사라지는데
- 이제 그거 말고 그룹 바이 
    - 아이템 아이디 기준으로 평균값

In [210]:
aaa.groupby('item_id').mean().sort_values('rating', ascending=False)[:5] #처음부터 5개 추천

NameError: name 'aaa' is not defined

In [211]:
aaa.merge(item)

NameError: name 'aaa' is not defined

In [216]:
aaa

NameError: name 'aaa' is not defined

In [249]:
aaa.groupby('movie_title').mean().sort_values('rating',ascending=False)#소팅하고

AttributeError: 'NoneType' object has no attribute 'groupby'

In [250]:
aaa.groupby('movie_title').mean().sort_values('rating',ascending=False)[:10]# 몇개 추천할건지. 처음부터 열개

AttributeError: 'NoneType' object has no attribute 'groupby'

In [ ]:
## 아까 머지랑 위에 그룹바이를 위 함수에 붙여넣어

### 인덱스를 칼럼으로 만드는거?
- unstack
    - 
- 세로가 가로 되려면? 시리즈 되야해
- 시리즈는 머지가 없어
    - 데이터 프레임으로 바꿔야해@@@@@
    - 

In [214]:
aaa.stack()

NameError: name 'aaa' is not defined

In [215]:
aaa.unstack()

NameError: name 'aaa' is not defined

In [ ]:
aaa.unstack().merge

In [217]:
pd.DataFrame(aaa.unstack())

NameError: name 'aaa' is not defined

# 여기에서 골치 아파
- item_id 에 대한 레이팅을 쭉 만들어...?!아까했..?
- 피봇?

In [ ]:
aaa.pivot()

In [ ]:
# 만약 더 큰애라면 그래서 계산이 많이 되는 애라면
- 저장하면 좋겠지?
- 텍스트가 아닌 객체 자체를 저장한다. 
- 객체 자체 저장하는 방법?

# 추천 패키지 있어요
- awesome java
- awesome python 등등
- 가져다 쓸 수 있지만 만드는거 회사가 더 좋아해 
- annoy, fastFM, implicit, libffm, lightfm,spotlight, Surprise, tensorrec